core/jvm/src/test/scala/fs2/ConcurrentlySpec.scala

```diff
@@ -60,31 +60,56 @@ class ConcurrentlySpec extends Fs2Spec with EventuallySupport {
 
     "run finalizers of background stream and properly handle exception" in forAll {
       s: PureStream[Int] =>
-        val prg = Stream
-          .eval(Deferred[IO, Unit])
-          .flatMap { halt =>
-            val bracketed =
-              Stream.bracket(IO(new java.util.concurrent.atomic.AtomicBoolean(true)))(
-                b => IO(b.set(false))
-              )
-
-            bracketed
-              .flatMap { b =>
-                s.get
-                  .covary[IO]
-                  .concurrently(
-                    (Stream.eval_(IO.sleep(50.millis)) ++
-                      Stream
-                        .eval_(halt.complete(())))
-                      .onFinalize(
-                        IO.sleep(100.millis) >>
-                          (if (b.get) IO.raiseError(new Err) else IO(()))
-                      ))
+        val Boom = new Err
+
+        val prg =
+          Ref.of[IO, Boolean](false).flatMap { runnerRun =>
+            Ref.of[IO, List[String]](Nil).flatMap { finRef =>
+              Deferred[IO, Unit].flatMap { halt =>
+                def bracketed: Stream[IO, Unit] =
+                  Stream.bracket(IO.unit)(
+                    _ => finRef.update(_ :+ "Outer")
+                  )
+
+                def runner: Stream[IO, Unit] =
+                  Stream
+                    .bracket(runnerRun.set(true))(_ =>
+                      IO.sleep(100.millis) >> // assure this inner finalizer always take longer run than `outer`
+                        finRef.update(_ :+ "Inner") >> // signal finalizer invoked
+                        IO.raiseError[Unit](Boom) // signal a failure
+                    ) >> // flag the concurrently had chance to start, as if the `s` will be empty `runner` may not be evaluated at all.
+                    Stream.eval_(halt.complete(())) // immediately interrupt the outer stream
+
+                val prg0 =
+                  bracketed
+                    .flatMap { b =>
+                      s.get.covary[IO].concurrently(runner)
+                    }
+                    .interruptWhen(halt.get.attempt)
+
+                prg0.compile.drain.attempt.flatMap { r =>
```

Comment:
Same comment here -- in general, most `.attempt.flatMap` calls should be replaced by `.guaranteeCase` or similar method from `Bracket[F]`.

Reply:
Yeah we could. I sort of used this patter, as this won't give us anything b/c we don't have special case for `cancel`. So was like sort of resistant to introduce another technique if we could live with attempt.flatMap.